ci-automation/garbage_collect.sh

```diff
@@ -38,26 +57,60 @@ function garbage_collect() {
 # --
 
 function _garbage_collect_impl() {
-    local keep="${1:-50}"
+    local keep="${1:-}"
+    local min_age_days="${2:-}"
     local dry_run="${DRY_RUN:-}"
     local purge_versions="${PURGE_VERSIONS:-}"
 
-    local versions_detected="$(git tag -l --sort=-committerdate \
-                | grep -E '(main|alpha|beta|stable|lts)-[0-9]+\.[0-9]+\.[0-9]+\-.*' \
-                | grep -vE '(-pro)$')"
+    # Set defaults; user-provided 'keep' has priority over default 'min_age_days'
+    if [ -n "${keep}" -a -z "${min_age_days}" ] ; then
+        min_age_days="0"
+    elif [ -z "${keep}" ] ; then
+        keep="50"
+    fi
+    if [ -z "${min_age_days}" ] ; then
+        min_age_days="14"
+    fi
 
-    echo "######## Full list of version(s) found ########"
-    echo "${versions_detected}" | awk '{printf "%5d %s\n", NR, $0}'
+    local min_age_date="$(date -d "${min_age_days} days ago" +'%Y-%m-%d')"
+    echo "######## Garbage collector starting ########"
+    echo
+    if [ -z "${purge_versions}" ] ; then
+        echo "Number of versions to keep: '${keep}'"
+        echo "Keep newer than: '${min_age_date}' (overrides number of versions to keep)"
+    fi
+    echo
 
     if [ -z "${purge_versions}" ] ; then
-        keep="$((keep + 1))" # for tail -n+...
+        # Generate a list "<timestamp> | <tagname>" from all repo tags that look like dev versions
+        local versions_detected="$(git tag -l --sort=-committerdate \
+                                          --format="%(creatordate:format:%Y-%m-%d) | %(refname:strip=2)" \
+                | grep -E '.*\| (main|alpha|beta|stable|lts)-[0-9]+\.[0-9]+\.[0-9]+-.*' \
+                | grep -vE '(-pro)$')"
+
+        echo "######## Full list of version(s) and their creation dates ########"
+        echo
+        echo "${versions_detected}" | awk '{printf "%5d %s\n", NR, $0}'
+
+        # Filter minimum number of versions to keep, min age
         purge_versions="$(echo "${versions_detected}" \
-                            | tail -n+"${keep}")"
+                            | awk -v keep="${keep}" -v min_age="${min_age_date}" '{
+                                if (keep > 0) {
+                                    keep = keep - 1
+                                    next
+                                }
+
+                                if ($1 > min_age)
+                                    next
+
+                                print $3
+                                }')"
     else
-        # make sure we only accept dev versions
+        # User-provided version list, make sure we only accept dev versions
         purge_versions="$(echo "${purge_versions}" | sed 's/ /\n/g' \
                             | grep -E '(main|alpha|beta|stable|lts)-[0-9]+\.[0-9]+\.[0-9]+\-.*' \
                             | grep -vE '(-pro)$')"
```

Comment:
Same here for excluding `-pro` from purges.

Reply:
We would need to rework tag removal and exclude any pattern that matches an official release version. Note that versions that look like official releases are skipped entirely - the garbage collector only cares about dev builds and nightlies. (Releases are supposed to be cleaned up manually, after pushing release binaries to the official mirrors.)